app/controllers/dashboard/dashboard-activity-controller.js

```diff
@@ -9,29 +9,49 @@ const moment = require('moment-timezone')
 const response = require('../../utils/response').response
 const CORRELATION_HEADER = require('../../utils/correlation_header').CORRELATION_HEADER
 const ConnectorClient = require('../../services/clients/connector_client').ConnectorClient
-const { isADirectDebitAccount } = require('../../services/clients/direct_debit_connector_client.js')
+const {isADirectDebitAccount} = require('../../services/clients/direct_debit_connector_client.js')
 const auth = require('../../services/auth_service.js')
 const connectorClient = () => new ConnectorClient(process.env.CONNECTOR_URL)
 const datetime = require('../../utils/nunjucks-filters/datetime')
 
 module.exports = (req, res) => {
   const correlationId = _.get(req, 'headers.' + CORRELATION_HEADER, '')
   const gatewayAccountId = auth.getCurrentGatewayAccountId((req))
+  let fromDateTime
   let toDateTime = moment().tz('Europe/London').format() // Today is the default
   let daysAgo = 0
-  let period = _.get(req, 'query.period', 'today')
+  const period = _.get(req, 'query.period', 'today')
 
   if (period === 'yesterday') {
     daysAgo = 1
   } else if (period === 'previous-seven-days') {
     daysAgo = 8 // 7+1 because we count starting from yesterday
   } else if (period === 'previous-thirty-days') {
     daysAgo = 31 // 30+1 because we count starting from yesterday
+  } else if (period === 'custom') {
+    fromDateTime = _.get(req, 'query.fromDateTime', null)
+    toDateTime = _.get(req, 'query.toDateTime', null)
+    if (!fromDateTime || !toDateTime || !moment(fromDateTime).isValid() || !moment(toDateTime).isValid()) {
+      logger.error(`[${correlationId}] Invalid custom date range specified`)
+      res.status(400)
+      response(req, res, 'dashboard/index', {
+        name: req.user.username,
+        serviceId: req.service.externalId,
+        activityError: true,
+        period
+      })
+      return
+    }
+    fromDateTime = decodeURIComponent(fromDateTime)
+    toDateTime = decodeURIComponent(toDateTime)
   }
-  const fromDateTime = moment().tz('Europe/London').startOf('day').subtract(daysAgo, 'days').format()
 
-  if (period !== 'today') {
-    toDateTime = moment().tz('Europe/London').startOf('day').format()
+  if (period !== 'custom') {
+    fromDateTime = moment().tz('Europe/London').startOf('day').subtract(daysAgo, 'days').format()
+
+    if (period !== 'today') {
+      toDateTime = moment().tz('Europe/London').startOf('day').format()
+    }
   }
```

Comment:
I think all this date logic needs to be extracted and unit tested. It's getting a bit convoluted